src/Adapter/MSTest.CoreAdapter/Execution/TestExecutionManager.cs

```diff
@@ -207,7 +222,103 @@ private void ExecuteTestsInSource(IEnumerable<TestCase> tests, IRunContext runCo
+                }
+
+                if (parallelLevel > 0 && sourceSettings.CanParallelizeAssembly)
+                {
+                    // Parallelization is enabled. Let's do further classification for sets.
+                    var logger = (IMessageLogger)frameworkHandle;
+                    logger.SendMessage(
+                        TestMessageLevel.Informational,
+                        string.Format(CultureInfo.CurrentCulture, Resource.TestParallelizationBanner, parallelLevel, parallelMode));
+
+                    // Create test sets for execution, we can execute them in parallel based on parallel settings
+                    IEnumerable<IGrouping<bool, TestCase>> testsets = Enumerable.Empty<IGrouping<bool, TestCase>>();
+
+                    // Parallel and not parallel sets.
+                    testsets = testsToRun.GroupBy(t => t.GetPropertyValue<bool>(TestAdapter.Constants.DoNotParallelizeProperty, false));
+
+                    var parallelizableTestSet = testsets.FirstOrDefault(g => g.Key == false);
+                    var nonparallelizableTestSet = testsets.FirstOrDefault(g => g.Key == true);
+
+                    if (parallelizableTestSet != null)
+                    {
+                        ConcurrentQueue<IEnumerable<TestCase>> queue = null;
+
+                        // Chunk the sets into further groups based on parallel level
+                        switch (parallelMode)
+                        {
+                            case TestParallelizationMode.MethodLevel:
+                                queue = new ConcurrentQueue<IEnumerable<TestCase>>(parallelizableTestSet.Select(t => new[] { t }).AsEnumerable<IEnumerable<TestCase>>());
+                                break;
+                            case TestParallelizationMode.ClassLevel:
+                                queue = new ConcurrentQueue<IEnumerable<TestCase>>(parallelizableTestSet.GroupBy(t => t.GetPropertyValue(TestAdapter.Constants.TestClassNameProperty) as string));
+                                break;
+                        }
+
+                        var tasks = new List<Task>();
+                        for (int i = 0; i < parallelLevel; i++)
+                        {
+                            tasks.Add(Task.Factory.StartNew(
+                                () =>
+                                {
+                                    while (!queue.IsEmpty)
+                                    {
+                                        if (queue.TryDequeue(out IEnumerable<TestCase> testSet))
+                                        {
+                                            this.ExecuteTestsWithTestRunner(testSet, runContext, frameworkHandle, source, sourceLevelParameters, testRunner);
+                                        }
+                                    }
+                                },
+                            CancellationToken.None,
```

Comment:
How does cancellation of a test run work for parallel? Should pass in `this.cancellationToken`?

Reply:
Before executing each test we check for a global Cancellation request from the platform, so we should be good.

Reply:
@AbhitejJohn In that case we will try to run all the tests even cancellation requested. Add check for global cancellation request while checking for queue.IsEmpty()?